presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java

```diff
@@ -140,11 +140,18 @@ public ScalarImplementation(
             return Optional.empty();
         }
         for (int i = 0; i < boundSignature.getArgumentTypes().size(); i++) {
-            Class<?> argumentType = typeManager.getType(boundSignature.getArgumentTypes().get(i)).getJavaType();
-            boolean nullableParameter = isParameterNullable(argumentType, nullableArguments.get(i), nullFlags.get(i));
-            Class<?> argumentContainerType = getNullAwareContainerType(argumentType, nullableParameter);
-            if (!argumentNativeContainerTypes.get(i).isAssignableFrom(argumentContainerType)) {
-                return Optional.empty();
+            if (boundSignature.getArgumentTypes().get(i).getBase().equals("function")) {
+                if (!lambdaInterface.containsKey(i) || !lambdaInterface.get(i).isAnnotationPresent(FunctionalInterface.class)) {
+                    return Optional.empty();
+                }
+            }
```

Comment:
The check of `lambdaInterface.get(i).isAnnotationPresent(FunctionalInterface.class)` should be in the constructor, not here.